drivers/flash/flash_ll_nrfxlib.c

```diff
@@ -0,0 +1,337 @@
+		    &flash_state.tmp_word,
+		    1,
+		    flash_operation_complete_callback);
+	}
+}
+
+static int flash_op_execute(void)
+{
+	int err;
+
+	err = MULTITHREADING_LOCK_ACQUIRE();
+	if (!err) {
+		if (flash_state.op == FLASH_OP_WRITE) {
+			err = flash_op_write();
+		} else if (flash_state.op == FLASH_OP_ERASE) {
+			flash_state.prev_len = NRF_FICR->CODEPAGESIZE;
+			err = ble_controller_flash_page_erase(
+			    (u32_t)flash_state.addr,
+			    flash_operation_complete_callback);
+		} else {
+			__ASSERT(0, "Unsupported operation\n");
+			err = -EINVAL;
+		}
+		MULTITHREADING_LOCK_RELEASE();
+	}
+	return err;
+}
+
+/* Driver API. */
+
+static int btctlr_flash_read(struct device *dev,
+			     off_t offset,
+			     void *data,
+			     size_t len)
+{
+	int err;
+
+	if (!is_addr_valid(offset, len)) {
+		return -EINVAL;
+	}
+
+	if (len == 0) {
+		return 0;
+	}
+
+	err = k_sem_take(&flash_state.sem, K_FOREVER);
+	if (err) {
+		return err;
+	}
+
+	memcpy(data, (void *)offset, len);
+	k_sem_give(&flash_state.sem);
+
+	return err;
+}
+
+static int btctlr_flash_write(struct device *dev,
+			      off_t offset,
+			      const void *data,
+			      size_t len)
+{
+	int err;
+
+	if (!is_addr_valid(offset, len)) {
+		return -EINVAL;
+	}
+
+	err = k_sem_take(&flash_state.sem, K_FOREVER);
+	if (err) {
+		return err;
+	}
+
+	__ASSERT_NO_MSG(flash_state.op == FLASH_OP_NONE);
+	flash_state.op = FLASH_OP_WRITE;
+	flash_state.data = data;
+	flash_state.addr = offset;
+	flash_state.len = len;
+
+	err = flash_op_execute();
+	if (err) {
+		k_sem_give(&flash_state.sem);
+	}
+
+	return err;
+}
+
+static int btctlr_flash_erase(struct device *dev, off_t offset, size_t len)
+{
+	int err;
+	int page_count = len / NRF_FICR->CODEPAGESIZE;
+
+	/* Follows the behavior of soc_flash_nrf.c */
+	if (!(is_page_aligned(offset) && is_page_aligned(len)) ||
+	    !is_addr_valid(offset, len)) {
+		return -EINVAL;
+	}
+
+	if (page_count == 0) {
+		return 0;
+	}
+
+	err = k_sem_take(&flash_state.sem, K_FOREVER);
+	if (err) {
+		return err;
+	}
+
+	__ASSERT_NO_MSG(flash_state.op == FLASH_OP_NONE);
+	flash_state.op = FLASH_OP_ERASE;
+	flash_state.addr = offset;
+	flash_state.len = len;
+
+	err = flash_op_execute();
+	if (err) {
+		k_sem_give(&flash_state.sem);
+	}
+
+	return err;
+}
+
+static int btctlr_flash_write_protection_set(struct device *dev, bool enable)
+{
+	/* The BLE controller handles the write protection automatically. */
+	return 0;
+}
+
+#if defined(CONFIG_FLASH_PAGE_LAYOUT)
+static struct flash_pages_layout dev_layout;
+
+static void btctlr_flash_page_layout_get(
+	struct device *dev,
+	const struct flash_pages_layout **layout,
+	size_t *layout_size)
+{
+	*layout = &dev_layout;
+	*layout_size = 1;
+}
+#endif /* defined(CONFIG_FLASH_PAGE_LAYOUT) */
+
+static int nrf_btctrl_flash_init(struct device *dev)
+{
+	dev->driver_api = &btctrl_flash_api;
+	k_sem_init(&flash_state.sem, 1, 1);
+
+#if defined(CONFIG_FLASH_PAGE_LAYOUT)
+	dev_layout.pages_count = NRF_FICR->CODESIZE;
```

Comment:
FICR is not available on SoC with trustzone in non-secure mode. This problem is covered by nrfx 1.7.1. I suggest to use it everywhere.

Reply:
The BLE controller only support the 52-series chips, so I don't think this is very relevant. However I can modify the code to use an nrfx API if you can point me to the right APIs.

Reply:
[uint32_t nrfx_nvmc_flash_size_get(void);](https://github.com/NordicPlayground/fw-nrfconnect-zephyr/blob/master/ext/hal/nordic/nrfx/drivers/include/nrfx_nvmc.h#L218)

[uint32_t nrfx_nvmc_flash_page_size_get(void);](https://github.com/NordicPlayground/fw-nrfconnect-zephyr/blob/master/ext/hal/nordic/nrfx/drivers/include/nrfx_nvmc.h#L225)

[uint32_t nrfx_nvmc_flash_page_count_get(void); ](https://github.com/NordicPlayground/fw-nrfconnect-zephyr/blob/master/ext/hal/nordic/nrfx/drivers/include/nrfx_nvmc.h#L232)